### PR TITLE
routing: assume TLV onion feature bit

### PR DESCRIPTION
**Pull request by ellemouton**
In this PR, we start assuming that hops on a path all support TLV onion encoding. 

This PR is the bare minimum. It does not completely remove the `Hop.LegacyPayload` member since we do still use this for deserialising persisted Hops for an HTLC attempt. So there is a chance that we have some on disk. And since we do have cases where we assume hops only support legacy encoding (either we did not receive their node announcement or we received chained hop hints in an invoice) the chances of this being the case are not that small. So some discussion/thought is perhaps needed around how we can handle this. 

The other thing we could potentially add to this PR is the removal of `--protocol.legacy.onion`

We already no longer read the `TlvPayload` boolean in `SendToRouteReq.Hop` so we could potentially remove this or warn the user if they are setting this to `false` that it is being ignored. 

Fixes: https://github.com/lightningnetwork/lnd/issues/8785

### CODE REVIEW

File: routing/pathfind_test.go [hunk 1]
```diff
@@ -1123,32 +1123,27 @@ func testBasicGraphPathFindingCase(t *testing.T, graphInstance *testGraphInstanc
...
+		payload, _, err := switchhop.ParseTLVPayload(
+			bytes.NewReader(sphinxPath[i].HopPayload.Payload),
+		)
+		require.NoError(t, err)
 
-		if !bytes.Equal(hopData.NextAddress[:], expectedHop[:]) {
-			t.Fatalf("first hop has incorrect next hop: expected %x, got %x",
-				expectedHop[:], hopData.NextAddress[:])
-		}
+		require.Equal(
```
Comment (bhandras) [inline routing/pathfind_test.go:1131]: super nitty: iiuc the expected value is the `route.Hops[i+1].ChannelID` so should be first. in `require.Equal`.

File: routing/pathfind_test.go [hunk 2]
```diff
@@ -1123,32 +1123,27 @@ func testBasicGraphPathFindingCase(t *testing.T, graphInstance *testGraphInstanc
...
+	)
+	require.NoError(t, err)
 
-	if !bytes.Equal(hopData.NextAddress[:], exitHop[:]) {
-		t.Fatalf("first hop has incorrect next hop: expected %x, got %x",
-			exitHop[:], hopData.NextAddress)
-	}
+	// The final hop should have a next hop value of all zeroes in order
+	// to indicate it's the exit hop.
+	require.EqualValues(t, payload.FwdInfo.NextHop.ToUint64(), 0)
```
Comment (bhandras) [inline routing/pathfind_test.go:1146]: nit: similarly to above comment :)

File: routing/pathfind_test.go [hunk 3]
```diff
@@ -1787,9 +1655,9 @@ func TestNewRoute(t *testing.T) {
 func runNewRoutePathTooLong(t *testing.T, useCache bool) {
 	var testChannels []*testChannel
 
-	// Setup a linear network of 21 hops.
+	// Setup a linear network of 26 hops.
```
Comment (bhandras) [inline routing/pathfind_test.go:1653]: 💯

**Review by bhandras — APPROVED** [review]
LGTM 🥇

File: routing/pathfind_test.go (hunk 1 above)
Comment (ellemouton) [reply routing/pathfind_test.go:1132]: indeedio

File: routing/pathfind_test.go (hunk 2 above)
Comment (ellemouton) [reply routing/pathfind_test.go]: changed it to require.Zero :)

**Review by ellemouton — COMMENTED** [review]
Thanks @bhandras ! 🥛

File: routing/pathfind_test.go [hunk 4]
```diff
@@ -1804,18 +1672,16 @@ func runNewRoutePathTooLong(t *testing.T, useCache bool) {
 
 	ctx := newPathFindingTestContext(t, useCache, testChannels, "start")
 
-	// Assert that we can find 20 hop routes.
-	node20 := ctx.keyFromAlias("node-20")
+	// Assert that we can find 25 hop routes.
+	node20 := ctx.keyFromAlias("node-25")
```
Comment (yyforyongyu) [inline routing/pathfind_test.go:1672]: nit: `node20` -> `node25`

File: routing/pathfind_test.go [hunk 5]
```diff
@@ -1804,18 +1672,16 @@ func runNewRoutePathTooLong(t *testing.T, useCache bool) {
...
 	payAmt := lnwire.MilliSatoshi(100001)
 	_, err := ctx.findPath(node20, payAmt)
 	require.NoError(t, err, "unexpected pathfinding failure")
 
-	// Assert that finding a 21 hop route fails.
-	node21 := ctx.keyFromAlias("node-21")
+	// Assert that finding a 26 hop route fails.
+	node21 := ctx.keyFromAlias("node-26")
```
Comment (yyforyongyu) [inline routing/pathfind_test.go:1678]: nit: `node21` -> `node26`

File: routing/route/route.go [hunk 6]
```diff
@@ -133,6 +133,11 @@ type Hop struct {
 	// LegacyPayload if true, then this signals that this node doesn't
 	// understand the new TLV payload, so we must instead use the legacy
 	// payload.
+	//
+	// NOTE: we should no longer ever create a Hop with Legacy set to true.
+	// The only reason we are keeping this member is that it could be the
+	// case that we have serialised hops persisted to disk where
+	// LegacyPayload is true.
```
Comment (yyforyongyu) [inline routing/route/route.go:140]: wonder when can we remove this completely🤔

**Review by yyforyongyu — APPROVED** [review]
Needs a rebase, otherwise LGTM🦾

File: routing/route/route.go (hunk 6 above)
Comment (ellemouton) [reply routing/route/route.go:140]: I _think_ we'd be able to do it quite soon. Afaict, we only serialise these for active payments & also for mission control. But for MC, we dont actually need this: so we should remove encoding all of this stuff for MC anyways and then for the payment scenario: as soon as no active payments have this anymore, we can remove it. So I think people just need to run this version for a bit (where no new legacy stuff gets persisted) and then after a few releases, we can just remove it.

**Review by ellemouton — COMMENTED** [review]
Thanks @yyforyongyu ! 🚀

**Comment by JssDWt** [comment]
Tested with the test setup described in https://github.com/lightningnetwork/lnd/issues/8785, it fixes the issue.

**Comment by ellemouton** [comment]
awesome - thanks for testing @JssDWt 🚀

**Pull request merged**